content/md/posts/clojure-cli-tools-understanding-aliases.md

```diff
@@ -0,0 +1,222 @@
+
+* `:paths` - a vector of directory names always included in the project class path, `["src" "resources"]`
+* `:deps` - a map of dependencies always included in the project ([practicalli/banking-on-clojure example](https://github.com/practicalli/banking-on-clojure-webapp/blob/live/deps.edn#L4-L19))
+* `:mvn/repos` - a map of repositories to download Maven dependencies from - Maven Central and Clojars are included by default in user wide deps.edn file ([practialli/clojure-deps-edn examples](https://github.com/practicalli/clojure-deps-edn/blob/live/deps.edn#L1140-L1161))
+* `:aliases` - a map of optional libraries and tools, the key being the alias name and its value the configuration ([practicalli/clojure-deps-edn examples](https://github.com/practicalli/clojure-deps-edn/blob/live/deps.edn#L42))
+
+
+## What is an alias?
+
+An alias is a way to add optional libraries or additional functionality to Clojure CLI tools, via a `deps.edn` configuration file.
+
+An alias name is a keyword in Clojure, e.g. `:env/test`, so the `:` is part of the alias name.
+
+As aliases are optional, they provide a way to add libraries and tools that are only needed for development and testing of Clojure projects.  For example, adding the `test` and `dev` directories to the Classpath, or test runners such as Kaocha.
+
+The configuration keys that can be used to define an alias are:
+
+* `extra-paths` - a vector of directory names always included in the project class path, `["dev" "test"]`
+* `:extra-deps` - a map of additional library dependencies
+* `:main-opts` - a vector of command line options passed to `clojure.main`
+* `exec-fn` - the fully qualified name of a function to be run by `clojure.exec`
+* `exec-args` - default arguments passed to the function, over-ridden by matching argument keys specified on the command line
+
+> Lesser used keys `:replace-paths` and `:replace-deps` use only the specified dependencies and directories, ignoring the project dependencies and class path. Community tools that use these directives should consider adopting the tool `-T` flag approach.
+
+
+Aliases are either added to the `deps.edn` file for a specific Clojure project, or to `$HOME/.clojure/deps.edn` (or `$XDG_HOME/.clojure.edn`) file for use with any Clojure project, as can be seen in [practicalli/clojure-deps-edn](https://github.com/practicalli/clojure-deps-edn).
+
+
+## Aliases defined in a project deps.edn
+
+A Clojure CLI tools project requires a `deps.edn` file in the root of the project directory. This typically contains `:path`, `:deps` and `:aliases` sections.
+
+A new project could be made by creating a `deps.edn` file and `src` & `test` directory trees.  The [clj-new project](https://github.com/seancorfield/clj-new) provides a convenient way to create a project from a template.
+
+The `:project/new` alias in [practicalli/clojure-deps-edn configuration](https://github.com/practicalli/clojure-deps-edn) contains a configuration for clj-new.
```

Comment:
I am still trying to figure out the details of using (-T) tools with the practicalli/clojure-deps-edn project.
I've created an up to date project using the latest clj-new (although I've removed the `:build` alias (not sure why an alias is needed if a `-Tbuild` tools should be used)
I am looking at deps-new as part of figuring out how best to incorprate that into this project.

Reply:
`clojure -Tfoo` runs a _tool_ that was installed `:as foo`. `clojure -T:bar` runs a tool that is added to the classpath via the alias `:bar`.

When you use `-T`, the `:paths` is set to `"."` (so it'll find `build.clj` in the root of a project) and the project dependencies, if any, are ignored (so only the "tool" is on the classpath).

So you run installed tools with, e.g., `clojure -Tclj-new ...` but you run the `build.clj` with `clojure -T:build ...` (typically, because the standard approach is to have a `:build` alias in your project `deps.edn` file).